### PR TITLE
Make more parallel get requests to ZooKeeper in system.zookeeper

### PR DESCRIPTION
**Pull request by davenger**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Followup for https://github.com/ClickHouse/ClickHouse/pull/51042 to make get requests more parallel as well.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit e1740edb67d145cbdaf0c3c33d350f14d1589169 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51118/e1740edb67d145cbdaf0c3c33d350f14d1589169/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by antonio2368** [comment]
Consider MultiRead for both `get` and `list`. E.g. you just send vector of paths to `get`/`tryGet`, if older Keeper version is detected same logic as you wrote will be applied (multiple parallel requests will be done). If Keeper supports MultiRead a single multi requests with `get` for all paths will be done.

**Comment by davenger** [comment]
Yep, I look into it. One thing that is going to be tricky is to handle failed requests with mutli-read batch. ASAIU in such a case all requests after the first failed one will also get errors. So In case of an error the rest of the batch needs to retried.

**Comment by antonio2368** [comment]
> ASAIU in such a case all requests after the first failed one will also get errors

No, it won't fail the entire Multi on the first failed subrequest because it makes it much less useful. Difference between `tryGet` and `get` is if it will throw on the first failed request from client side.

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 1]
```diff
@@ -459,75 +460,93 @@ void ReadFromSystemZooKeeper::fillData(MutableColumns & res_columns) const
 
             task.path_corrected = pathCorrected(path);
 
-            task.future = zookeeper->asyncTryGetChildren(task.path_corrected);
+            list_requests.emplace_back(zkutil::makeListRequest(task.path_corrected));
             list_tasks.emplace_back(std::move(task));
         }
+        Coordination::Responses list_responses;
+        zookeeper->tryMulti(list_requests, list_responses);
```
Comment (antonio2368) [inline src/Storages/System/StorageSystemZooKeeper.cpp:467]: why not `tryGetChildren` and

Comment (davenger) [reply src/Storages/System/StorageSystemZooKeeper.cpp:467]: > why not `tryGetChildren` and

```throw DB::Exception(ErrorCodes::CANNOT_PARSE_TEXT, "Not enough information in comment")``` :)

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 2]
```diff
@@ -459,75 +460,92 @@ void ReadFromSystemZooKeeper::fillData(MutableColumns & res_columns) const
...
-                context->getProcessListElement()->checkTimeLimit();
-                auto res = futures[i].get();
-                if (res.error == Coordination::Error::ZNONODE)
-                    continue; /// Node was deleted meanwhile.
-
-                // Deduplication
-                String key = path_part + '/' + nodes[i];
-                if (auto [it, inserted] = added.emplace(key); !inserted)
-                    continue;
+                get_requests.emplace_back(zkutil::makeGetRequest(task.path_part + '/' + node));
```
Comment (antonio2368) [inline src/Storages/System/StorageSystemZooKeeper.cpp:504]: you can do the same with `tryGet` 😄

**Review by antonio2368 — APPROVED** [review]

File: src/Storages/System/StorageSystemZooKeeper.cpp [hunk 3]
```diff
@@ -459,75 +460,91 @@ void ReadFromSystemZooKeeper::fillData(MutableColumns & res_columns) const
...
+
+        for (size_t i = 0, size = get_tasks.size(); i < size; ++i)
+        {
+            auto & res = get_responses[i];
+            if (res.error == Coordination::Error::ZNONODE)
+                continue; /// Node was deleted meanwhile.
+
+            auto & get_task = get_tasks[i];
+            auto & list_task = list_tasks[get_task.list_task_idx];
+            context->getProcessListElement()->checkTimeLimit();
```
Comment (tavplubix) [inline src/Storages/System/StorageSystemZooKeeper.cpp:519]: It's possible to run it from a background operation, so `context->getProcessListElement()` will be nullptr

Comment (tavplubix) [reply src/Storages/System/StorageSystemZooKeeper.cpp:519]: https://pastila.nl/?00170ec1/4e423038e8e2d0f12b2806aa3910e38b#+nBHJqZPSOArPKUFv7LIMw==

**Pull request merged**